include/fiction/algorithms/simulation/sidb/ground_state_space.hpp

```diff
@@ -0,0 +1,896 @@
+        // and set that of the parent; what the new parent determines what interactions are external
+        for (const sidb_cluster_ptr& child : parent->children)
+        {
+            for (const uint64_t sidb_ix : child->sidbs)
+            {
+                const sidb_cluster_receptor_state child_rst{child, sidb_ix};
+                subtract_sibling_pot_from_received_ext_pot_bound<bound_direction::LOWER>(parent, child_rst);
+                subtract_sibling_pot_from_received_ext_pot_bound<bound_direction::UPPER>(parent, child_rst);
+            }
+        }
+    }
+
+    bool verify_composition(sidb_charge_space_composition& composition) const noexcept
+    {
+        // initialize the composition potential bounds to (0.0, 0.0) for all SiDBs
+        composition.pot_bounds.initialise_complete_potential_bounds(top_cluster->num_sidbs());
+
+        // perform physically informed space pruning for a multiset composition
+        for (sidb_cluster_projector_state& receiving_pst : composition.proj_states)
+        {
+            for (const uint64_t sidb_ix : receiving_pst.cluster->sidbs)
+            {
+                double internal_pot_lb{};
+                double internal_pot_ub{};
+
+                for (const sidb_cluster_projector_state& pst : composition.proj_states)
+                {
+                    internal_pot_lb += get_projector_state_bound<bound_direction::LOWER>(pst, sidb_ix).pot_val;
+                    internal_pot_ub += get_projector_state_bound<bound_direction::UPPER>(pst, sidb_ix).pot_val;
+                }
+
+                composition.pot_bounds.set(sidb_ix, internal_pot_lb, internal_pot_ub);
+            }
+
+            if (!perform_potential_bound_analysis<potential_bound_analysis_mode::ANALYZE_COMPOSITION>(
+                    receiving_pst, composition.pot_bounds))
+            {
+                return false;
+            }
+        }
+
+        return true;
+    }
+
+    void fill_merged_charge_state_space(const sidb_cluster_ptr& parent, const uint64_t cur_child_ix,
+                                        sidb_cluster_charge_state& m) const noexcept
+    {
+        if (cur_child_ix >= parent->children.size())
+        {
+            if (!verify_composition(m.compositions.front()))
+            {
+                return;
+            }
+
+            // check if cluster charge state exists
+            const auto it = parent->charge_space.find(m);
+            if (it != parent->charge_space.cend())
+            {
+                it->compositions.emplace_back(m.compositions.front());
+            }
+            else
+            {
+                parent->charge_space.emplace(m);
+            }
+
+            return;
+        }
+
+        const sidb_cluster_ptr& cur_child = *std::next(parent->children.cbegin(), static_cast<int64_t>(cur_child_ix));
+
+        for (const sidb_cluster_charge_state& m_part : cur_child->charge_space)
+        {
+            m.compositions.front().proj_states.emplace_back(
```

Comment:
Can we always be sure that ``m.compositions`` is not empty? We may get undefined behavior otherwise.

Reply:
Yes, it will always be filled, since I use the `.front()` entry to store all possible charge space compositions in.